agreement/player.go

```diff
@@ -260,6 +275,64 @@ func (p *player) handleCheckpointEvent(r routerHandle, e checkpointEvent) []acti
 		}}
 }
 
+func (p *player) handleWinningPayloadArrival(payload proposal, ver protocol.ConsensusVersion) {
+	// ignoring validatedAt
+	p.payloadArrivals = append(p.payloadArrivals, payload.receivedAt)
+	p.resizePayloadArrivals(ver)
+}
+
+func (p *player) resizePayloadArrivals(ver protocol.ConsensusVersion) {
+	proto := config.Consensus[ver]
+	if len(p.payloadArrivals) > proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = p.payloadArrivals[len(p.payloadArrivals)-proto.DynamicFilterPayloadArriavalHistory:]
+	}
+	for len(p.payloadArrivals) < proto.DynamicFilterPayloadArriavalHistory {
+		p.payloadArrivals = append([]time.Duration{FilterTimeout(0, ver)}, p.payloadArrivals...)
+	}
+}
+
+// calculateFilterTimeout chooses the appropriate filter timeout.
+func (p *player) calculateFilterTimeout(ver protocol.ConsensusVersion, tracer *tracer) time.Duration {
+
+	proto := config.Consensus[ver]
+
+	if !proto.DynamicFilterTimeout || p.Period != 0 {
+		// Either dynamic lambda is disabled, or we're not in period 0 and
+		// therefore, can't use dynamic lambda
+		return FilterTimeout(p.Period, ver)
+	}
+
+	var dynamicDelay time.Duration
+	defaultDelay := FilterTimeout(0, ver)
+	if proto.DynamicFilterPayloadArriavalHistory <= 0 {
+		// we don't keep any history, use the default
+		dynamicDelay = defaultDelay
+	} else if proto.DynamicFilterPayloadArriavalHistory > len(p.payloadArrivals) {
+		// not enough samples, use the default
+		dynamicDelay = defaultDelay
+	} else {
+		sortedArrivals := make([]time.Duration, len(p.payloadArrivals))
+		copy(sortedArrivals[:], p.payloadArrivals[:])
+		sort.Slice(sortedArrivals, func(i, j int) bool { return sortedArrivals[i] < sortedArrivals[j] })
+		dynamicDelay = sortedArrivals[proto.DynamicFilterPayloadArriavalHistory-1]
```

Comment:
Is there a world where we adjust our heuristic for the available data? For example, using an exponential weighted moving average? Potentially a single outlier data point here spikes our round times for up to `DynamicFilterPayloadArrivalHistory` rounds until that falls out of the cache.

Also, iterating here and keeping track of the highest value seen is faster than sorting--probably negligible for such small data sizes though.

Reply:
I changed this logic to look at the 95th percentile now. 
The current code supports changing this percentile or the history window size through parameters but not the algorithm (e.g., to exponential moving average).